test/topology/test_change_ip.py

```diff
@@ -8,56 +8,118 @@
+                    logger.info(f"len(system.local) == {len(local_rows)} != 1")
+                    return None
+                local_row = local_rows[0]
+                local_vals = (local_row.host_id, local_row.listen_address, local_row.rpc_address, local_row.broadcast_address)
+                local_expected_vals = (uuid.UUID(host_id), ip, ip, ip)
+                if local_vals != local_expected_vals:
+                    logger.info(f"local vals {local_vals} doesn't match expected vals {local_expected_vals}")
+                    return None
+
+                current_peers = {r.host_id: r.peer for r in await manager.cql.run_async("select peer, host_id from system.peers", host=host)}
+                if current_peers != expected_peers:
+                    logger.info(f"host {host} current peers {current_peers} doesn't match the expected_peers {expected_peers}, continue waiting")
+                    return None
+
+                alive_eps = set(await manager.api.get_alive_endpoints(ip))
+                if alive_eps != expected_alive_ips:
+                    logger.info(f"host {host} alive endpoints {alive_eps} doesn't match the expected alive ips {expected_alive_ips}, continue waiting")
+                    return None
+
+                down_eps = set(await manager.api.get_down_endpoints(ip))
+                if down_eps != expected_down_ips:
+                    logger.info(f"host {host} down endpoints {down_eps} doesn't match the expected down ips {expected_down_ips}, continue waiting")
+                    return None
+
+                actual_host_id_map = {x['value']: x['key'] for x in await manager.api.get_host_id_map(host.address)}
+                if expected_host_id_map != actual_host_id_map:
+                    logger.info(f"host {host} id map {actual_host_id_map} doesn't match the expected host id map {expected_host_id_map}, continue waiting")
+                    return None
+
+                return True
+
+            await wait_for(see_proper_ips, time.time() + 60)
+
+    # We're checking the crash scenario here - the servers[0] crashes just after
+    # saving s1_new_ip but before removing s1_old_ip. After its restart we should
+    # see s1_new_ip.
+    if mode != 'release':
+        await manager.api.enable_injection(servers[0].ip_addr, 'crash-before-prev-ip-removed', one_shot=True)
+        await manager.api.enable_injection(servers[0].ip_addr, 'ip-change-raft-sync-delay', one_shot=False)
+    await manager.server_start(servers[1].server_id)
+    servers[1] = ServerInfo(servers[1].server_id, s1_new_ip, s1_new_ip)
+    if mode != 'release':
+        s0_logs = await manager.server_open_log(servers[0].server_id)
+        await s0_logs.wait_for('crash-before-prev-ip-removed hit, killing the node')
+        await manager.server_stop(servers[0].server_id)
+        await manager.server_start(servers[0].server_id)
+        await manager.api.enable_injection(servers[0].ip_addr, 'ip-change-raft-sync-delay', one_shot=False)
```

Comment:
Depends on which notification this is supposed to catch -- if it's supposed to observe `servers[1]` or itself on restart, that notification may happen before you inject, so the injection will not work. If you're injecting this to catch the `servers[2]` start below, then it's fine.

(BTW. comments would be helpful, why we're delaying these notifications and which notifications we want to delay)

Reply:
> (BTW. comments would be helpful, why we're delaying these notifications and which notifications we want to delay)

This is mentioned in the commit message.
> Note that we need to inject an additional delay ip-change-raft-sync-delay
> before old IP is removed. Otherwise the problem stop reproducing - other
> nodes remove the old IP before it's send back to the just restarted node.

> Depends on which notification this is supposed to catch -- if it's supposed to observe servers[1] or itself on restart, that notification may happen before you inject, so the injection will not work. (And then you should use the startup injections.) If you're injecting this to catch the servers[2] start below, then it's fine.

Injection into `servers[0]` should catch `server[1]` and `server[2]` restart. The `server[0]` itself restarts `if mode != 'release'`, so we need to inject `ip-change-raft-sync-delay` again after restart.

Reply:
> This is mentioned in the commit message.

I meant comments in the test code. The commit message is much harder to access (you need to `git blame` etc.)